### PR TITLE
Feature/additional functions

### PR DESCRIPTION
**Pull request by musangowope**
## Proposed changes

- Add debounce function with doc block explainations
- Added hasValue function  with doc blocks explainations and tests
- added isObjectEmpty functions with js doc block explainations and tests
- added moveItemInArray function with js doc block and test
- added stringFormatting function with doc blocks and test

## Types of changes

What type of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New feature
- [ ] Bugfix
- [ ] Quality of Life

## Checklist

_Put an `x` in the boxes that apply._

- [x] Linter passes (`yarn lint`)
- [x] Unit Test pass (`yarn jest`)
- [ ] I have added at least two useful tests for new component (if applicable)
- [ ] I have added at least one Storybook story (if applicable)

## Relevant Issues
resolves https://github.com/bonlineza/b-lib/issues/66

### CODE REVIEW

File: src/util/debounced.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/**
+ * function that enforces programming practice used to ensure that
+ * time-consuming task do not fire so often, that it stalls the performance
+ * of the web page. Limiting the rate at which a function gets invoked.
+ *
+ *
+ * @return function
+ * @param delay
```
Comment (sscholle) [inline src/util/debounced.js]: please define the param TYPE and DESCRIPTION like: `@param delay {number} the number of milliseconds to delay the function call`

File: src/util/hasValue.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Function that accepts data and determines whether that data has
+ * value or not
+ *
+ *
+ * @return boolean
+ * @param value
+ */
+
```
Comment (sscholle) [inline src/util/hasValue.js]: docblocks should not have breaks after, it must be 'attached' to the function it describes

File: src/util/moveItemInArray.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Shifts element array to designated position in array. fromIndex
+ * is the chosen array element and toIndex is the array position for
+ * you want the element to go
+ * @return array
+ * @param array
```
Comment (sscholle) [inline src/util/moveItemInArray.js]: please ensure all PARAM blocks have TYPES and DESCRIPTIONS

File: src/util/debounced.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+/**
+ * function that enforces programming practice used to ensure that
+ * time-consuming task do not fire so often, that it stalls the performance
+ * of the web page. Limiting the rate at which a function gets invoked.
```
Comment (robguy21) [inline src/util/debounced.js]: this is a very wordy way of saying "Ensures a function is only run only once every `delay`" 😅

Comment (robguy21) [reply src/util/debounced.js]: which is also a wordy way of saying "Debounces a function for `delay` ms"

File: src/util/hasValue.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Function that accepts data and determines whether that data has
+ * value or not
```
Comment (robguy21) [inline src/util/hasValue.js]: this isn't correct, `undefined`, `null`, and `''` are all valid values in javascript- we should rather say in the description that it ensures that the value passed in is not one of {insert values we do not want to consider}

File: src/util/isObjectEmpty.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Function that accepts object as parameter and checks whether
+ * Object is empty or not
+ * @return boolean
+ * @param {object} obj: the object that is checked to see whether it has
+ * any keys
```
Comment (robguy21) [inline src/util/isObjectEmpty.js]: i think `@param`  must come before `@return`

File: src/util/moveItemInArray.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Shifts element array to designated position in array. fromIndex
+ * is the chosen array element and toIndex is the array position for
+ * you want the element to go
+ * @return array
+ * @param {array} array: array for which items will be shifted
+ * @param {number} fromIndex: index of array item that is going to move
+ * @param {number} toIndex: position for which the chosen index item is going to
+ * shift to
```
Comment (robguy21) [inline src/util/moveItemInArray.js]: `@return` should come last

File: src/util/moveItemInArray.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Shifts element array to designated position in array. fromIndex
+ * is the chosen array element and toIndex is the array position for
+ * you want the element to go
```
Comment (robguy21) [inline src/util/moveItemInArray.js]: "Shifts element *in an* array to *a* designated position."

The rest of this description should be the `@param` descriptions

File: src/util/debounced.js (hunk 4 above)
Comment (musangowope) [reply src/util/debounced.js]: done

File: src/util/isObjectEmpty.js (hunk 6 above)
Comment (musangowope) [reply src/util/isObjectEmpty.js]: done

File: src/util/hasValue.js (hunk 5 above)
Comment (musangowope) [reply src/util/hasValue.js]: done

File: src/util/moveItemInArray.js (hunk 7 above)
Comment (musangowope) [reply src/util/moveItemInArray.js]: done

File: src/util/moveItemInArray.js (hunk 8 above)
Comment (musangowope) [reply src/util/moveItemInArray.js]: done

File: src/util/debounced.js (hunk 1 above)
Comment (musangowope) [reply src/util/debounced.js]: done

File: src/util/hasValue.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+/**
+ * ensures that the value passed in is not null or undefined
```
Comment (robguy21) [inline src/util/hasValue.js]: sorry to be pedantic but it also ensures that it is not an empty string

File: src/util/moveItemInArray.js (hunk 8 above)
Comment (robguy21) [reply src/util/moveItemInArray.js]: > fromIndex
> is the chosen array element and toIndex is the array position for
> you want the element to go

That can all be removed since you're describing the parameters here, no need to describe parameters in the description of the function

File: src/util/stringFormatting.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Function that accepts string and replaces underscores with spaces
+ *
+ * @return string
+ * @param {string} str: the string that the transformation happens on
+ * turning underscore into spaces
+ */
```
Comment (robguy21) [inline src/util/stringFormatting.js]: `@return` after `@param`

File: src/util/stringFormatting.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+export function replaceUnderscoreWithSpace(str) {
+  return str.replace(/_/g, ' ');
+}
+/**
+ * Function that accepts string and capitalizes first character of every word
+ *
+ * @return string
+ * @param {string} str: the first character of each word is capitalized for
+ * this string
+ */
```
Comment (robguy21) [inline src/util/stringFormatting.js]: `@return` after `@param`

File: src/util/stringFormatting.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+/**
+ * Function that accepts string and replaces underscores with spaces and
+ * capitalizes (the first character of every word uppercased)
+ *
+ * @return boolean
+ * @param {string} str: string for which transformation happens for capitalizes
+ * string
+ */
```
Comment (robguy21) [inline src/util/stringFormatting.js]: `@return` after `@param`

File: src/util/stringFormatting.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+export function toTitleCase(str) {
+  return str.replace(
+    /\w\S*/g,
+    txt => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase(),
+  );
+}
+
+/**
+ * Function that accepts string and replaces underscores with spaces and
+ * capitalizes (the first character of every word uppercased)
```
Comment (robguy21) [inline src/util/stringFormatting.js]: why is 
> the first character of every word uppercased

surrounded by `(` `)` ?

File: src/util/isObjectEmpty.js [hunk 14]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Function that accepts object as parameter and checks whether
+ * Object is empty or not
+ * @param {object} obj: the object that is checked to see whether it has
+ * @return boolean
+ * any keys
```
Comment (sscholle) [inline src/util/isObjectEmpty.js]: whats this?

File: src/util/moveItemInArray.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Shifts element array to designated position in array. fromIndex
+ * is the chosen array element and toIndex is the array position for
```
Comment (sscholle) [inline src/util/moveItemInArray.js]: this is already written in the param description...

File: src/util/stringFormatting.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Function that accepts string and replaces underscores with spaces
+ *
```
Comment (sscholle) [inline src/util/stringFormatting.js]: no need to have a space after the function description

File: src/util/stringFormatting.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+ * turning underscore into spaces
+ */
+export function replaceUnderscoreWithSpace(str) {
+  return str.replace(/_/g, ' ');
+}
+/**
+ * Function that accepts string and capitalizes first character of every word
+ *
+ * @return string
+ * @param {string} str: the first character of each word is capitalized for
```
Comment (sscholle) [inline src/util/stringFormatting.js]: watch your ordering

File: src/util/hasValue.js (hunk 9 above)
Comment (musangowope) [reply src/util/hasValue.js]: done

File: src/util/stringFormatting.js (hunk 10 above)
Comment (musangowope) [reply src/util/stringFormatting.js]: done

File: src/util/stringFormatting.js (hunk 11 above)
Comment (musangowope) [reply src/util/stringFormatting.js]: done

File: src/util/stringFormatting.js (hunk 12 above)
Comment (musangowope) [reply src/util/stringFormatting.js]: doen

File: src/util/stringFormatting.js (hunk 13 above)
Comment (musangowope) [reply src/util/stringFormatting.js]: refactored

File: src/util/isObjectEmpty.js (hunk 14 above)
Comment (musangowope) [reply src/util/isObjectEmpty.js]: reordered to make sense

File: src/util/moveItemInArray.js (hunk 15 above)
Comment (musangowope) [reply src/util/moveItemInArray.js]: removed

File: src/util/stringFormatting.js (hunk 16 above)
Comment (musangowope) [reply src/util/stringFormatting.js]: done

File: src/util/stringFormatting.js (hunk 17 above)
Comment (musangowope) [reply src/util/stringFormatting.js]: reordered

File: src/util/isObjectEmpty.js [hunk 18]
```diff
@@ -0,0 +1,6 @@
+/**
+ * @param {object} obj: the object that is checked to see whether it has
```
Comment (sscholle) [inline src/util/isObjectEmpty.js]: would be nice to have a description

File: src/util/isObjectEmpty.spec.js [hunk 19]
```diff
@@ -0,0 +1,11 @@
+import isObjectEmpty from './isObjectEmpty';
+
+describe('isObjectEmpty', () => {
+  it('function isObjectEmpty with "{}" as a parameter should return true', () => {
+    expect(isObjectEmpty({})).toBe(true);
+  });
+
+  it('function isObjectEmpty with "{ isSad: false }" as a parameter should return true', () => {
```
Comment (sscholle) [inline src/util/isObjectEmpty.spec.js]: isSad is sad

File: src/util/moveItemInArray.js [hunk 20]
```diff
@@ -0,0 +1,23 @@
+/**
+ * @param {array} array: array for which items will be shifted
```
Comment (sscholle) [inline src/util/moveItemInArray.js]: no function description?

File: src/util/stringFormatting.js [hunk 21]
```diff
@@ -0,0 +1,27 @@
+/**
+ * @param {string} str: the underscores of this string turn into spaces
```
Comment (sscholle) [inline src/util/stringFormatting.js]: what happened to the all the function descriptions?

File: src/util/hasValue.js [hunk 22]
```diff
@@ -0,0 +1,7 @@
+/**
+ * ensures that the value passed in is not null or undefined or is not an
+ * empty
```
Comment (robguy21) [inline src/util/hasValue.js]: an empty... string :)

File: src/util/isObjectEmpty.js [hunk 23]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Function that accepts object as parameter and checks whether
+ * Object is empty or not
```
Comment (robguy21) [inline src/util/isObjectEmpty.js]: or not what though?

File: src/util/moveItemInArray.js [hunk 24]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Shifts element array to designated position in array.
```
Comment (robguy21) [inline src/util/moveItemInArray.js]: 🎉

File: src/util/isObjectEmpty.js [hunk 25]
```diff
@@ -0,0 +1,8 @@
...
+ * @param {object} obj: the object that is checked to see whether it has
```
Comment (sscholle) [inline src/util/isObjectEmpty.js]: 1. `{object}` is no a valid type
2. `Function that accepts object` is already described by the params
3. a more concise description could be: `Checks whether an Object has no props`

File: src/util/moveItemInArray.js (hunk 24 above)
Comment (sscholle) [inline src/util/moveItemInArray.js]: good!

File: src/util/smoothScrollTo.js [hunk 26]
```diff
@@ -77,7 +77,6 @@ export const scrollToElement = element => {
     }
   }
 };
-
```
Comment (sscholle) [inline src/util/smoothScrollTo.js]: rather keep that space for readability

File: src/util/stringFormatting.js [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+export function toTitleCase(str) {
+  return str.replace(
+    /\w\S*/g,
+    txt => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase(),
+  );
+}
+
+/**
+ * Function that accepts string then removes underscores and capitalizes
+ * @param {string} str: string which undergoes two modifications 1)
```
Comment (robguy21) [inline src/util/stringFormatting.js]: "string to modify" is fine here, you're again re-describing the function in the parameter description...

> two modifications 1)
> replacing underscores with spaces and capitalizing the first character of
> each word

Is kind of covered with 

> Function that accepts string then removes underscores and capitalizes

So all of that can be nuked

File: src/util/stringFormatting.js [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+ * each word
+ * @return string
+ */
+export function snakeCaseToTitleCase(str) {
+  return toTitleCase(replaceUnderscoreWithSpace(str));
+}
+/**
+ * Function that accepts string as parameter and returns string without
+ * white spaces
+ * @param {string} str: string for which the white space have been removed
```
Comment (robguy21) [inline src/util/stringFormatting.js]: again describing what the function does in the parameter description

File: src/util/stringFormatting.js [hunk 29]
```diff
@@ -0,0 +1,41 @@
+/**
+ * Function that accepts string and replaces underscores with spaces
```
Comment (sscholle) [inline src/util/stringFormatting.js]: 1. please don't start describing the function based on what it *accepts*: this is evident by the params
2. example: `Replace all underscore string ocurrances with space, ie: '_' -> ' '

File: src/util/stringFormatting.js [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+ */
+export function toTitleCase(str) {
+  return str.replace(
+    /\w\S*/g,
+    txt => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase(),
+  );
+}
+
+/**
+ * Function that accepts string then removes underscores and capitalizes
```
Comment (sscholle) [inline src/util/stringFormatting.js]: BAD!

File: src/util/hasValue.js [hunk 31]
```diff
@@ -0,0 +1,7 @@
+/**
+ * ensures that the value passed in is not null or undefined or is not an empty
```
Comment (sscholle) [inline src/util/hasValue.js]: 1. i'm not sure `ensures` is the correct wording it does not ensure anything. it only 'checks'
2. again we are taking about the 'value passed' this is the same as 'accepts'
3. use: 'Checks if a value is not `null`, `undefined` or an empty string'

File: src/util/isObjectEmpty.js [hunk 32]
```diff
@@ -0,0 +1,7 @@
+/**
+ * Determines whether the object is empty by checking whether the length
+ * of key for that object is false
```
Comment (sscholle) [inline src/util/isObjectEmpty.js]: 1. describing how the function works internally is not required - ie: `the length of keys` is an implementation detail and very wordy, `props` describes it just as well and much shorter, please use minimum of descriptive words

Comment (sscholle) [reply src/util/isObjectEmpty.js]: 2. rather use: `Checks if a value has any props`

File: src/util/smoothScrollTo.js (hunk 26 above)
Comment (sscholle) [reply src/util/smoothScrollTo.js]: note this

File: src/util/stringFormatting.js [hunk 33]
```diff
@@ -0,0 +1,37 @@
+/**
+ * Replaces underscores with spaces
+ * @param {string} str: string to modify
+ * @return string
+ */
+export function replaceUnderscoreWithSpace(str) {
+  return str.replace(/_/g, ' ');
+}
+/**
```
Comment (sscholle) [inline src/util/stringFormatting.js]: space above comment please

File: src/util/stringFormatting.js [hunk 34]
```diff
@@ -0,0 +1,37 @@
...
+ * @param {string} str: string to modify
+ * @return string
+ */
+export function toTitleCase(str) {
+  return str.replace(
+    /\w\S*/g,
+    txt => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase(),
+  );
+}
+
```
Comment (sscholle) [inline src/util/stringFormatting.js]: good!

File: src/util/stringFormatting.js [hunk 35]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Replaces underscores with spaces and capitalizes the first character of
+ * each word
+ * @param {string} str: string to modify
+ * @return string
+ */
+export function snakeCaseToTitleCase(str) {
+  return toTitleCase(replaceUnderscoreWithSpace(str));
+}
+/**
```
Comment (sscholle) [inline src/util/stringFormatting.js]: space

File: src/util/debounced.js [hunk 36]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Ensures a function is only run only once every delay
+ * @param delay {number} the number of milliseconds to delay the function call
+ * @return function
```
Comment (robguy21) [inline src/util/debounced.js]: this should be `@returns {Function}` i believe https://jsdoc.app/tags-returns.html

File: src/util/hasValue.js [hunk 37]
```diff
@@ -0,0 +1,7 @@
+/**
+ * Checks if a value is not null, undefined or an empty string
+ * string
+ * @param {any} value: this checked to see whether it has a value
```
Comment (sscholle) [inline src/util/hasValue.js]: `the value to be checked`

File: src/util/moveItemInArray.js [hunk 38]
```diff
@@ -0,0 +1,24 @@
...
+ * @param {array} array: array for which items will be shifted
+ * @param {number} fromIndex: index of array item that is going to move
+ * @param {number} toIndex: Designated position in array for which the element
```
Comment (sscholle) [inline src/util/moveItemInArray.js]: `...array to which...`

File: src/util/moveItemInArray.js [hunk 39]
```diff
@@ -0,0 +1,24 @@
...
+ * is shifted to
+ * @returns array
```
Comment (sscholle) [inline src/util/moveItemInArray.js]: type needs braces

Comment (sscholle) [reply src/util/moveItemInArray.js]: return needs description: ie `the new rearranged array`

File: src/util/smoothScrollTo.js (hunk 26 above)
Comment (sscholle) [reply src/util/smoothScrollTo.js]: where is that space?

File: src/util/stringFormatting.js [hunk 40]
```diff
@@ -0,0 +1,39 @@
...
+ * Replaces underscores with spaces
+ * @param {string} str: string to modify
+ * @returns string
```
Comment (sscholle) [inline src/util/stringFormatting.js]: braces, description

File: src/util/debounced.js [hunk 41]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Creates closure for delaying function
+ * @param delay {number} the number of milliseconds to delay the function call
```
Comment (robguy21) [inline src/util/debounced.js]:
```
@param delay {number} 
```

should be

```
@param {number} delay
```

File: src/util/validateEmail.js [hunk 42]
```diff
@@ -6,7 +6,9 @@ const EmailTest = new RegExp(`^${user}@(${domain}|(\\[(${ipv4}|${ipv6})\\]))$`);
 
 /**
  * Checks if the given string matches good known Email validation patterns
- * @param {string} value the string to be tested against acceptable email formatting rules
+ * @param {string} value the string to be tested against acceptable
```
Comment (sscholle) [inline src/util/validateEmail.js]: too wordy (duplication)

File: src/util/validatePhone.js [hunk 43]
```diff
@@ -1,8 +1,9 @@
 const PhoneTest = /^\+(\d{11})$/;
 
 /**
- * Checks wether the given string value matches international telephomne formatting
+ * Checks wether the given string value matches international telephone formatting
  * @param {string} value the international formatted telephone number to be validated
```
Comment (sscholle) [inline src/util/validatePhone.js]: too wordy: `the phone number to be validated`
